specs/HubProtocol.md

```diff
@@ -0,0 +1,225 @@
+    "invocationId": 123,
+    "target": "Send",
+    "arguments": [
+        42,
+        "Test Message"
+    ]
+}
+```
+
+### Result Message Encoding
+
+A `Result` message is a JSON object with the following properties:
+
+* `type` - A `Number` with the literal value 2, indicating that this message is a Result.
+* `invocationId` - A `String` encoding the `Invocation ID` for a message.
+* `error` - A `String` encoding the error message. The presence of this field indicates that an error should be yielded, **even if the payload field is also present**
+* `payload` - A `Token` encoding the result value (see "JSON Payload Encoding" for details). If error is present, this field is **completely ignored**.
+
+Example - A `Result` with a payload.
+
+```json
+{
+    "type": 2,
+    "invocationId": 123,
+    "payload": 42
+}
+```
+
+Example - A `Result` with an error.
+
+```json
+{
+    "type": 2,
+    "invocationId": 123,
+    "error": "An error occurred!"
+}
+```
+
+Example - Even though there is a `payload`, the presence of `error` means it is ignored and the error is yielded. This should never be transmitted by a conforming endpoint, but endpoints should be prepared to receive messages of this format, to be robust.
+
+```json
+{
+    "type": 2,
+    "invocationId": 123,
+    "error": "An error occurred!",
+    "payload": 42
+}
+```
+
+### Completion Message Encoding
+
+A `Completion` message is a JSON object with the following properties
+
+* `type` - A `Number` with the literal value `3`, indicating that this message is a `Completion`.
+* `invocationId` - A `String` encoding the `Invocation ID` for a message.
+
+Example
+
+```json
+{
+    "type": 3,
+    "invocationId": 123
+}
+```
+
+### JSON Payload Encoding
+
+Items in the arguments array within the `Invocation` message type, as well as the payload value of the `Result` message, encode values which have meaning to each particular Binder. A general guideline for encoding/decoding these values is provided in the "Type Mapping" section at the end of this document, but Binders should provide configuration to applications to allow them to customize these mappings. These mappings need not be self-describing, because when decoding the value, the Binder is expected to know the destination type (by looking up the definition of the method indicated by the Target).
+
+## Protocol Buffers (ProtoBuf) Encoding
+
+In order to support ProtoBuf, an application must provide a [ProtoBuf service definition](https://developers.google.com/protocol-buffers/docs/proto3) for the Hub. However, implementations may automatically generate these definitions from reflection information, if the underlying platform supports this. For example, the .NET implementation will attempt to generate service definitions for methods that use only simple primitive and enumerated types. The service definition provides a description of how to encode the arguments and return value for the call. For example, consider the following C# method:
```

Comment:
I am missing details about how things should work with types not supported out of the box.